Add pre-auth driver to allow/reject per user connection

rndb8418 opened the pull request:
Hello,

This is in relation to: https://github.com/drakkan/sftpgo/pull/826

This adds the ability to allow/reject a connection once the `USER` command has been received by adding an interface the server driver can implement. The driver can then decide whether to allow or reject the connection based on the connection & username it is supplied.

For example, it may want to allow insecure/plaintext FTP for specific users and require TLS (FTPS) for everyone else. This extension will allow SFTPGo to have per-user configuration on how user connections are handled, before they transmit their full authentication details.

drakkan commented on driver.go:
```diff
@@ -50,6 +50,14 @@ type MainDriverExtensionPassiveWrapper interface {
+	// PreAuthUser is called when receiving the "USER" command before proceeding with any other checks
+	// If it returns true, the user is allowed to proceed to authentication assuming it passes other checks
+	// If it returns false or a non-nil error, the client will receive a 530 error and be disconnected.
+	PreAuthUser(cc ClientContext, user string) (bool, error)
```
I think returning an error here is enough, we really need a boolean too?

drakkan commented on handle_auth.go:
```diff
@@ -13,6 +13,24 @@ func (c *clientHandler) handleUSER(param string) error {
+
+			return nil
+		}
+	}
```
basically we return the same message `if err != nil` and `if !result`. I think the boolean is not unnecessary. I would also change the error message from `Authentication problem: %v", err` to `User not allowed: %v", err`, `User rejected` or something similar

rndb8418 replied:
Do you think we should perhaps allow the message to be fully customised by the implementation then they can specify exactly the message they want to return without a hard coded part included? i.e. just `"%v"` which would allow implementations to use wording that isn't English.

drakkan replied:
> Do you think we should allow the message to be fully customised by the implementation then they can specify exactly the message they want to return without a hard coded part included? i.e. just `"%v"` which would allow implementations to use wording that isn't English.

I think we should add this feature in a separate PR and make all other messages translatable as well. I'm not sure about the best approach to localize a Go app

drakkan reviewed (changes requested):
I would also add a setting on the clientHandler to enforce a particular TLS requirement so that we can control both the global setting and this new per client setting and then enforce the TLS requirement on transfer connections as well. Maybe this could be done in a separate commit but in SFTPGo I don't like to have a global setting that checks both control and data connections and a per-user setting that only checks control connection

rndb8418 commented:
@drakkan thanks, I've simplified the return to error only. I will look at the data connection enforcement too as it makes sense to be able to hook into both.

drakkan commented:
> @drakkan thanks, I've simplified the return to error only. I will look at the data connection enforcement too as it makes sense to be able to hook into both.

ok, thanks please also fix the linting issue, take a look [here](https://github.com/fclairamb/ftpserverlib/runs/6465961273?check_suite_focus=true). 

Sorry if I'm not very responsive here, but I have my day job and some paid SFTPGo related consultancy. I have practically no time

rndb8418 commented:
Thanks, understand your time is limited and your responsiveness is fine,

I've renamed the code to better suite the purpose (rather than just a pre-auth check), as it now from your suggestion also handles the data connections. I did attempt adding this to the `TransferOpen` initially but it meant the data channel request (from `PASV`/`PORT` etc.) was returning a success response to client which I think might be less ideal as it's better to reject as soon as possible vs. waiting until they try to do a transfer?

Adding it to `TransferOpen` would mean more simple checks can pass (such as mandatory TLS) prior to invoking a call that could be more detailed on a per user/connection basis, and reduces the repetition that I've added into the active/passive areas, but depends on whether we want to reject early as possible or not.

I believe most of the lint issues are now fixed too, though there's one that crops up around the `handlePASV` method due to it's length/number of statements.

rndb8418 commented on client_handler.go:
```diff
@@ -667,6 +667,14 @@ func (c *clientHandler) checkDataConnectionRequirement(dataConnIP net.IP, channe
 	}
 }
 
+func (c *clientHandler) checkDataConnectionHook() error {
```
That sounds like a simpler approach. Is there any other use cases where we might want to allow customisation as this would potentially cover other use cases for the libraries?

As a made up example, if you wanted to allow a user to connect at any time, but only allow data connections between 10AM and 12PM, this extension would allow that logic to be supported by specifying it in the implementing library. Though functionality for a client is quite limited without a data connection (can't even list directories etc.), so not sure if there is many use cases for this much extensibility.

drakkan replied:
I'm not sure. I generally take a look at estabilished alternative to understand what the users wants, see for example here

http://www.proftpd.org/docs/howto/TLS.html

when I added the global TLS requirement configuration I used an enum to allow to add other requirements. Then I thought that options like `auth` or `auth+data` were there for historical reasons and I didn't added them. I don't see a time restriction there but as said above your implementation can be ok, let's see what @fclairamb thinks about it. I really have no time for this now, I have to return to my daily job. Thanks

drakkan replied:
In any case, the TLS Requirement check must be done in TransferOpen, as suggested above. With your patch, an attacker can steal a port and open an unencrypted transfer connection

rndb8418 replied:
Happy to move that, should be able to submit the updated code today/tomorrow.

When you have a moment to clarify, can you explain how the move to `TransferOpen` protects against an attacker stealing a port/opening an unencrypted data connection?

I can see how this could be possible now in a `PASV` connection as the listener [here](https://github.com/fclairamb/ftpserverlib/blob/6098c14d1a0e7247a7fa049a8eb34f1d12d76672/transfer_pasv.go#L128-L132) is established prior to the check I've added. However it seems like this would still be an issue when moving the check to `TransferOpen` as that only gets called once a command like `LIST`, `STOR` etc. is called and the data connection is already established as that would happen after the `PASV` command completed?

rndb8418 replied:
Thanks for explanation, it makes sense to keep the checks in a consistent location. I have moved the logic there now and can swap to a property check once confirmed that's the direction that is wanted.

I think in a future PR, it might be worth migrating the both checks to where the listener is established (`handlePASV` and `handlePORT`) to enable it being prevented earlier on (at start of their request for connection, rather than afterwards). I think having the checks done twice (in both the `handlePASV`/`handlePORT` + `TransferOpen` would be redundant unless they are able to clear the TLS once the data connection is already established (I don't think that's the case but might be wrong!).

fclairamb replied:
Hi @rndb8418 & @drakkan,
sorry for the late answer.

I think both approaches make sense. In my opinion:
- @drakkan's idea with `SetTLSRquirement` seems closer to how developers will probably use it
- @rndb8418's current driver extension has the big advantage of not adding complexity to the `clientHandler`

So, I think the current implementation is reasonable is pretty good 👍

fclairamb reviewed (approved):
LGTM

drakkan commented:
@fclairamb thanks. It looks good for me too, but please leave me a few more days to try out a test implementation in SFTPGo (June 2 is a holiday in Italy, I should be able to find some time), thanks

drakkan commented:
David, `DataConnectionAllowed` does not pass the username. With the current patch I think there is no way to add what are you asking for to SFTPGo. Am I wrong?

If we add the username to `DataConnectionAllowed`, in SFTPGo we have to query the data provider every time a data connection is requested and this will degrade performance or add some specific cache.

On the SFTPGo side it seems easier to set the TLS requirement on the ClientContext at login time and let ftpserverlib check the TLS requirement, but my point of view may be biased.

@fclairamb If you think this is the best solution I do not object to merging this patch as is or with minor changes (for example adding the username).

rndb8418 commented:
@drakkan my understanding is that the `CommandConnectionAllowed` requires the username to be passed, as it's not yet specified on the `clientHandler` instance, which occurs here, after the check: https://github.com/fclairamb/ftpserverlib/blob/dfee3af174339924566cef5ea13349ca832786f3/handle_auth.go#L50

I had thought that the `DataConnectionAllowed` did not need this, as it is passed the `clientHandler` which has a `user` property, however I now realise that's not accessible in the `DataConnectionAllowed` call as it receives the implementation of interface `ClientContext` and doesn't provide access to the user property.

I think it's fine to add username as second parameter then if it can't be accessed, or a `User()` method could be added to `ClientContext`to return the username (implemented via `clientHandler`) if present?

drakkan commented:
I have no preference, you can add the user to `DataConnectionAllowed` or the exported interface. 
But please note that for SFTPGo I will not accept a patch that queries the data provider every time an FTP user opens a data connection.

drakkan commented:
Fixed in #334

rndb8418 commented:
@drakkan thank you, your solution seems eloquent and I now see it's advantages of having it set once per connection rather than it needing to check multiple times. I don't think I understood that properly before.

Appreciate your time on it, and sorry I hadn't found the time to complete this PR but I think that worked out for the better!
